File: packages/sdk-flutter/lib/src/models/sdk_settings.dart
/// SDK Settings model for controlling API tracing and logging behavior
class NivoStackSdkSettings {
  // Security settings
  final bool captureRequestBodies;
  final bool captureResponseBodies;
  final bool capturePrintStatements;
  final bool sanitizeSensitiveData;
  final List<String> sensitiveFieldPatterns;

  // Performance settings
  final int maxLogQueueSize;
  final int maxTraceQueueSize;
  final int flushIntervalSeconds;
  final bool enableBatching;

  // Log control
  final String minLogLevel;
  final bool verboseErrors;

  const NivoStackSdkSettings({
    this.captureRequestBodies = true,
    this.captureResponseBodies = true,
    this.capturePrintStatements = false,
    this.sanitizeSensitiveData = true,
    this.sensitiveFieldPatterns = const ['password', 'token', 'secret', 'apiKey', 'api_key', 'authorization', 'cookie'],
    this.maxLogQueueSize = 100,
    this.maxTraceQueueSize = 50,
    this.flushIntervalSeconds = 30,
    this.enableBatching = true,
    this.minLogLevel = 'debug',
    this.verboseErrors = false,
  });

  /// Create default settings
  factory NivoStackSdkSettings.defaults() => const NivoStackSdkSettings();

  /// Create from JSON response
  factory NivoStackSdkSettings.fromJson(Map<String, dynamic> json) {
    return NivoStackSdkSettings(
      captureRequestBodies: json['captureRequestBodies'] ?? true,
      captureResponseBodies: json['captureResponseBodies'] ?? true,
      capturePrintStatements: json['capturePrintStatements'] ?? false,
      sanitizeSensitiveData: json['sanitizeSensitiveData'] ?? true,
      sensitiveFieldPatterns: json['sensitiveFieldPatterns'] != null
          ? List<String>.from(json['sensitiveFieldPatterns'])
          : const ['password', 'token', 'secret', 'apiKey', 'api_key', 'authorization', 'cookie'],
      maxLogQueueSize: json['maxLogQueueSize'] ?? 100,
      maxTraceQueueSize: json['maxTraceQueueSize'] ?? 50,
      flushIntervalSeconds: json['flushIntervalSeconds'] ?? 30,
      enableBatching: json['enableBatching'] ?? true,
      minLogLevel: json['minLogLevel'] ?? 'debug',
      verboseErrors: json['verboseErrors'] ?? false,
    );
  }

  /// Check if a log level should be captured based on minLogLevel
  bool shouldCaptureLogLevel(String level) {
    // If logging is disabled, don't capture any logs
    if (minLogLevel.toLowerCase() == 'disabled') {
      return false;
    }

    const levelOrder = ['verbose', 'debug', 'info', 'warn', 'error', 'assert'];
    final minIndex = levelOrder.indexOf(minLogLevel.toLowerCase());
    final levelIndex = levelOrder.indexOf(level.toLowerCase());

    // If either level is unknown, default to capturing (unless disabled)
    if (minIndex == -1 || levelIndex == -1) {
      return minLogLevel.toLowerCase() != 'disabled';
    }

    return levelIndex >= minIndex;
  }

  /// Sanitize a string value by redacting sensitive patterns
  String sanitizeValue(String key, String value) {
    if (!sanitizeSensitiveData) return value;

    final keyLower = key.toLowerCase();
    for (final pattern in sensitiveFieldPatterns) {
      if (keyLower.contains(pattern.toLowerCase())) {
        return '[REDACTED]';
      }
    }
    return value;
  }

  /// Sanitize a map by redacting sensitive fields
  Map<String, dynamic> sanitizeMap(Map<String, dynamic> data) {
    if (!sanitizeSensitiveData) return data;

    return data.map((key, value) {
      if (value is String) {
        return MapEntry(key, sanitizeValue(key, value));
      } else if (value is Map<String, dynamic>) {
        return MapEntry(key, sanitizeMap(value));
      }
      return MapEntry(key, value);
    });
  }

  @override
  String toString() {
    return 'NivoStackSdkSettings('
        'captureRequestBodies: $captureRequestBodies, '
        'captureResponseBodies: $captureResponseBodies, '
        'capturePrintStatements: $capturePrintStatements, '
        'sanitizeSensitiveData: $sanitizeSensitiveData, '
        'sensitiveFieldPatterns: $sensitiveFieldPatterns, '
        'maxLogQueueSize: $maxLogQueueSize, '
        'maxTraceQueueSize: $maxTraceQueueSize, '
        'flushIntervalSeconds: $flushIntervalSeconds, '
        'enableBatching: $enableBatching, '
        'minLogLevel: $minLogLevel, '
        'verboseErrors: $verboseErrors'
        ')';
  }
}

/// API endpoint configuration for per-endpoint settings
class NivoStackApiConfig {
  final String endpoint;
  final String? method;
  final bool enableLogs;
  final bool captureRequestBody;
  final bool captureResponseBody;
  final double? costPerRequest;

  const NivoStackApiConfig({
    required this.endpoint,
    this.method,
    this.enableLogs = true,
    this.captureRequestBody = true,
    this.captureResponseBody = true,
    this.costPerRequest,
  });

  factory NivoStackApiConfig.fromJson(Map<String, dynamic> json) {
    return NivoStackApiConfig(
      endpoint: json['endpoint'] ?? '',
      method: json['method'],
      enableLogs: json['enableLogs'] ?? true,
      captureRequestBody: json['captureRequestBody'] ?? true,
      captureResponseBody: json['captureResponseBody'] ?? true,
      costPerRequest: json['costPerRequest']?.toDouble(),
    );
  }

  /// Check if this config matches a given URL and method
  bool matches(String url, String requestMethod) {
    // Extract path from URL
    String path;
    try {
      final uri = Uri.parse(url);
      path = uri.path;
    } catch (e) {
      path = url;
    }

    // Check if path matches endpoint pattern
    if (!_pathMatches(path, endpoint)) return false;

    // Check method if specified
    if (method != null && method!.isNotEmpty && method!.toUpperCase() != requestMethod.toUpperCase()) {
      return false;
    }

    return true;
  }

  bool _pathMatches(String path, String pattern) {
    // Simple path matching - check if path starts with or equals pattern
    // Remove trailing slashes for comparison
    final normalizedPath = path.endsWith('/') ? path.substring(0, path.length - 1) : path;
    final normalizedPattern = pattern.endsWith('/') ? pattern.substring(0, pattern.length - 1) : pattern;

    return normalizedPath == normalizedPattern || normalizedPath.startsWith('$normalizedPattern/');
  }
}

File: packages/sdk-ios/Sources/NivoStack/Models/SdkSettings.swift
import Foundation

/// SDK Settings model for controlling API tracing and logging behavior
public struct NivoStackSdkSettings: Codable {
    // Security settings
    public let captureRequestBodies: Bool
    public let captureResponseBodies: Bool
    public let capturePrintStatements: Bool
    public let sanitizeSensitiveData: Bool
    public let sensitiveFieldPatterns: [String]
    
    // Performance settings
    public let maxLogQueueSize: Int
    public let maxTraceQueueSize: Int
    public let flushIntervalSeconds: Int
    public let enableBatching: Bool
    
    // Log control
    public let minLogLevel: String
    public let verboseErrors: Bool
    
    public init(
        captureRequestBodies: Bool = true,
        captureResponseBodies: Bool = true,
        capturePrintStatements: Bool = false,
        sanitizeSensitiveData: Bool = true,
        sensitiveFieldPatterns: [String] = ["password", "token", "secret", "apiKey", "api_key", "authorization", "cookie"],
        maxLogQueueSize: Int = 100,
        maxTraceQueueSize: Int = 50,
        flushIntervalSeconds: Int = 30,
        enableBatching: Bool = true,
        minLogLevel: String = "debug",
        verboseErrors: Bool = false
    ) {
        self.captureRequestBodies = captureRequestBodies
        self.captureResponseBodies = captureResponseBodies
        self.capturePrintStatements = capturePrintStatements
        self.sanitizeSensitiveData = sanitizeSensitiveData
        self.sensitiveFieldPatterns = sensitiveFieldPatterns
        self.maxLogQueueSize = maxLogQueueSize
        self.maxTraceQueueSize = maxTraceQueueSize
        self.flushIntervalSeconds = flushIntervalSeconds
        self.enableBatching = enableBatching
        self.minLogLevel = minLogLevel
        self.verboseErrors = verboseErrors
    }
    
    /// Create default settings
    public static func defaults() -> NivoStackSdkSettings {
        return NivoStackSdkSettings()
    }
    
    /// Create from JSON dictionary
    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        captureRequestBodies = try container.decodeIfPresent(Bool.self, forKey: .captureRequestBodies) ?? true
        captureResponseBodies = try container.decodeIfPresent(Bool.self, forKey: .captureResponseBodies) ?? true
        capturePrintStatements = try container.decodeIfPresent(Bool.self, forKey: .capturePrintStatements) ?? false
        sanitizeSensitiveData = try container.decodeIfPresent(Bool.self, forKey: .sanitizeSensitiveData) ?? true
        sensitiveFieldPatterns = try container.decodeIfPresent([String].self, forKey: .sensitiveFieldPatterns) ?? ["password", "token", "secret", "apiKey", "api_key", "authorization", "cookie"]
        maxLogQueueSize = try container.decodeIfPresent(Int.self, forKey: .maxLogQueueSize) ?? 100
        maxTraceQueueSize = try container.decodeIfPresent(Int.self, forKey: .maxTraceQueueSize) ?? 50
        flushIntervalSeconds = try container.decodeIfPresent(Int.self, forKey: .flushIntervalSeconds) ?? 30
        enableBatching = try container.decodeIfPresent(Bool.self, forKey: .enableBatching) ?? true
        minLogLevel = try container.decodeIfPresent(String.self, forKey: .minLogLevel) ?? "debug"
        verboseErrors = try container.decodeIfPresent(Bool.self, forKey: .verboseErrors) ?? false
    }
    
    enum CodingKeys: String, CodingKey {
        case captureRequestBodies
        case captureResponseBodies
        case capturePrintStatements
        case sanitizeSensitiveData
        case sensitiveFieldPatterns
        case maxLogQueueSize
        case maxTraceQueueSize
        case flushIntervalSeconds
        case enableBatching
        case minLogLevel
        case verboseErrors
    }
    
    /// Check if a log level should be captured based on minLogLevel
    public func shouldCaptureLogLevel(_ level: String) -> Bool {
        // If logging is disabled, don't capture any logs
        if minLogLevel.lowercased() == "disabled" {
            return false
        }
        
        let levelOrder = ["verbose", "debug", "info", "warn", "error", "assert"]
        guard let minIndex = levelOrder.firstIndex(where: { $0 == minLogLevel.lowercased() }),
              let levelIndex = levelOrder.firstIndex(where: { $0 == level.lowercased() }) else {
            // If either level is unknown, default to capturing (unless disabled)
            return minLogLevel.lowercased() != "disabled"
        }
        
        return levelIndex >= minIndex
    }
    
    /// Sanitize a string value by redacting sensitive patterns
    public func sanitizeValue(key: String, value: String) -> String {
        if !sanitizeSensitiveData { return value }
        
        let keyLower = key.lowercased()
        for pattern in sensitiveFieldPatterns {
            if keyLower.contains(pattern.lowercased()) {
                return "[REDACTED]"
            }
        }
        return value
    }
    
    /// Sanitize a dictionary by redacting sensitive fields
    public func sanitizeMap(_ data: [String: Any]) -> [String: Any] {
        if !sanitizeSensitiveData { return data }
        
        var sanitized: [String: Any] = [:]
        for (key, value) in data {
            if let stringValue = value as? String {
                sanitized[key] = sanitizeValue(key: key, value: stringValue)
            } else if let dictValue = value as? [String: Any] {
                sanitized[key] = sanitizeMap(dictValue)
            } else {
                sanitized[key] = value
            }
        }
        return sanitized
    }
}

/// API endpoint configuration for per-endpoint settings
public struct NivoStackApiConfig: Codable {
    public let endpoint: String
    public let method: String?
    public let enableLogs: Bool
    public let captureRequestBody: Bool
    public let captureResponseBody: Bool
    public let costPerRequest: Double?
    
    public init(
        endpoint: String,
        method: String? = nil,
        enableLogs: Bool = true,
        captureRequestBody: Bool = true,
        captureResponseBody: Bool = true,
        costPerRequest: Double? = nil
    ) {
        self.endpoint = endpoint
        self.method = method
        self.enableLogs = enableLogs
        self.captureRequestBody = captureRequestBody
        self.captureResponseBody = captureResponseBody
        self.costPerRequest = costPerRequest
    }
    
    /// Check if this config matches a given URL and method
    public func matches(url: String, requestMethod: String) -> Bool {
        // Extract path from URL
        let path: String
        if let uri = URL(string: url) {
            path = uri.path
        } else {
            path = url
        }
        
        // Check if path matches endpoint pattern
        if !pathMatches(path: path, pattern: endpoint) { return false }
        
        // Check method if specified
        if let method = method, !method.isEmpty {
            if method.uppercased() != requestMethod.uppercased() {
                return false
            }
        }
        
        return true
    }
    
    private func pathMatches(path: String, pattern: String) -> Bool {
        // Simple path matching - check if path starts with or equals pattern
        // Remove trailing slashes for comparison
        let normalizedPath = path.hasSuffix("/") ? String(path.dropLast()) : path
        let normalizedPattern = pattern.hasSuffix("/") ? String(pattern.dropLast()) : pattern
        
        return normalizedPath == normalizedPattern || normalizedPath.hasPrefix("\(normalizedPattern)/")
    }
}

